### PR TITLE
Updated CamelCase to snake_case converter to cover more cases

### LINKED ISSUE #406: Acronyms in tablename aren't converted to snake case
Acronyms aren't being recognized as separate words when converting to snake case.  Only happens with plain acronyms or acronyms with a prefix but not a suffix.

``` python
from flask import Flask
from flask_sqlalchemy import SQLAlchemy

app = Flask(__name__)
db = SQLAlchemy(app)

class ABC(db.Model):
    id = db.Column(db.Integer, primary_key=True)

class PreABC(db.Model):
    id = db.Column(db.Integer, primary_key=True)

class ABCPost(db.Model):
    id = db.Column(db.Integer, primary_key=True)

class PreABCPost(db.Model):
    id = db.Column(db.Integer, primary_key=True)

db.engine.echo = True
db.create_all()
```

```
CREATE TABLE "ABC" (
    id INTEGER NOT NULL, 
    PRIMARY KEY (id)
)

CREATE TABLE "preABC" (
    id INTEGER NOT NULL, 
    PRIMARY KEY (id)
)

CREATE TABLE abc_post (
    id INTEGER NOT NULL, 
    PRIMARY KEY (id)
)

CREATE TABLE pre_abc_post (
    id INTEGER NOT NULL, 
    PRIMARY KEY (id)
)
```

### PR DESCRIPTION
**Pull request by Saif807380**
I have updated the regex pattern for converting CamelCase table names to snake_case. I've also compared this with #617. Detailed results of the comparison can be found [here](http://bit.ly/test-case-analysis). 

`tests/test_basic_app.py::test_sqlite_relative_path` fails for `SQLAlchemy == 1.4.0` giving #885. All tests pass with older `SQLAlchemy` version.

- fixes #406 

Checklist:

- [x] Add tests that demonstrate the correct behavior of the change. Tests should fail without the change.
- [x] Add or update relevant docs, in the docs folder and in code.
- [x] Add an entry in `CHANGES.rst` summarizing the change and linking to the issue.
- [x] Add `.. versionchanged::` entries in any relevant code docs.
- [x] Run `pre-commit` hooks and fix any issues.
- [x] Run `pytest` and `tox`, no tests failed.

### CODE REVIEW

**Comment by davidism** [comment]
Found a few more test cases I had written out a long time ago. There's probably some overlap with what you tested, but you can add others.

```python
@pytest.mark.parametrize(
    ("name", "expect"),
    [
        ("A", "a"),
        ("A1", "a1"),
        ("1A", "1a"),
        ("CamelCase", "camel_case"),
        ("Snake_Case", "snake_case"),
        ("ABC", "abc"),
        ("HTMLLayout", "html_layout"),
        ("LayoutHTML", "layout_html"),
        ("HTTP2Request", "http2_request"),
        ("4XTest", "4x_test"),
        ("ShoppingCartSession", "shopping_cart_session"),
    ],
)
```

**Comment by davidism** [comment]
Also, use `@pytest.mark.parametrize()` as shown above. This allows each test case to pass/fail individually.

File: tests/test_table_name.py [hunk 1]
```diff
@@ -4,6 +4,54 @@
...
+        "vp9_codec",
+        "http2_request_session",
+        "user_st4",
+        "http2_client_type3_encoder_parameters_sse",
+        "long_name4_testing_camel_case2snake_case_kek",
+        "foo_bar_sse2",
+        "alarm_message_ss2_signal_transformer",
+        "ast_v2_node",
+        "http_response_code_xyz",
+        "get_htt_presponse_code",
```
Comment (ThiefMaster) [inline tests/test_table_name.py:45]: YUCK ;) but I guess we can't really avoid such edge cases

**Comment by Saif807380** [comment]
I'll update the test

File: tests/test_table_name.py (hunk 1 above)
Comment (davidism) [reply tests/test_table_name.py:45]: To be fair, the original name was yuck too.
